internal/customdiags/with_error_code.go

```diff
@@ -0,0 +1,63 @@
+package customdiags
+
+import (
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+// ErrorDiagnosticWithErrorCode is an error diagnostic that stored the error code.
+type ErrorDiagnosticWithErrorCode struct {
```

Comment:
Two comments:
1. Can we be explicit that this is the HTTPStatusCode. Since there is both the gRPC Status Code and the HTTP Status Code, it is slightly ambiguous what code to check.
2. Since we are in a package called `customdiags`, what do you think about shortening the struct name to something like `ErrorHTTPStatusCode` since it will be imported as `customdiags.ErrorHTTPStatusCode`